### PR TITLE
Add a Lucene 9.8 engine.

### LINKED ISSUE #44: Lucene 9.6 engine actually uses Lucene 8.10.1?
I have been playing with this benchmark lately to better understand where Lucene had room for improvement compared to other engines (thank you!) and noticed that the `lucene-9.6` engine actually had a dependency on Lucene 8.10.1: https://github.com/quickwit-oss/search-benchmark-game/blob/0ca432c568abbdf95f74e0d95c5652748fd0d6c7/engines/lucene-9.6.0/build.gradle#L15-L17.

I plan on contributing a `lucene-9.8` engine soo, I could remove this `lucene-9.6` engine at that time if this looks like a good way forward.

### PR DESCRIPTION
**Pull request by jpountz**
This PR does a few changes:
 - Removes the lucene-9.6.0 engine which is actually built on 8.10.1 rather than 9.6.0.
 - Adds a lucene-9.8.0 engine.
 - Reuses the original Lucene JARs instead of creating an uber JAR.
 - Makes indexing multi-threaded so that it takes less time to build the index (just for convenience since indexing is not measured).
 - Adds document reordering to speed up query processing, especially conjunctions.
 - TOP_10 no longer counts up to 1,000 hits, it stops at 10.

Closes #44

### CODE REVIEW

File: engines/lucene-9.8.0/build.gradle [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+version '1.0-SNAPSHOT'
+
+sourceCompatibility = 17
+
+repositories {
+    mavenCentral()
+    maven {
+        url "https://dist.apache.org/repos/dist/dev/lucene/lucene-9.8.0-RC1-rev-d914b3722bd5b8ef31ccf7e8ddc638a87fd648db/lucene/maven/"
```
Comment (jpountz) [inline engines/lucene-9.8.0/build.gradle:12]: This bit needs to be removed before merging the PR, in order to switch from the temporary Maven repo that is used for voting on the Lucene 9.8 release (https://lists.apache.org/thread/sg6n375511k6zf31g7qbv0fc4rsvjb89) to the actual Lucene 9.8 release.

Comment (jpountz) [reply engines/lucene-9.8.0/build.gradle]: This is now removed.

**Comment by jpountz** [comment]
Lucene 9.8 is released, this PR is now ready from my point of view.

**Comment by PSeitz** [comment]
Thanks! How does the index reordering work?

**Comment by jpountz** [comment]
This implements an algorithm called "Recursive graph bisection", which is the same algorithm that the `pisa-0.8.2` engine uses. Here is the Lucene PR where it was implemented: https://github.com/apache/lucene/pull/12489.

This algorithm tries to reorder doc IDs to minimize the sum of the logs of gaps between consecutive doc IDs in postings. This helps make postings more compact, but this also has the interesting side-effect of clustering documents that have similar postings together, which in-turn helps conjunctions.

The [original paper](https://arxiv.org/pdf/1602.08820.pdf) is good but I found this [follow-up reproducibility study](http://engineering.nyu.edu/~suel/papers/bp-ecir19.pdf) to be a better introduction to this algorithm.

**Comment by PSeitz** [comment]
That's pretty interesting. I think it would be nice to make this configurable

**Comment by jpountz** [comment]
Do you have opinions on how to make this configurable? FWIW testing performance without reordering is as simple as replacing `idx-reordered` with `idx` in the Makefile target that runs queries. If we want to make it easier to test with and without doc ID reordering, I think I could easily split the Lucene 9.8 engine into `lucene-9.8` and `lucene-9.8-bp` or something like that.

**Comment by PSeitz** [comment]
For now I think it would be good to have `lucene-9.8` and `lucene-9.8-reorder`

**Comment by jpountz** [comment]
@PSeitz It's now split, I used the name `lucene-9.8.0-bp` if you don't mind to be more explicit about what reordering is implemented (e.g. another option is to sort the index by a min_hash of the terms of each document).

I'm not too familiar with `make` so there might be better ways of doing what I tried to do...

**Comment by PSeitz** [comment]
Thanks! There may be better ways with `make`, but the duplication variant is unamibigious

**Pull request merged**